zilliqa/src/api/eth.rs

```diff
@@ -816,6 +766,377 @@ fn get_uncle(_: Params, _: &Arc<Mutex<Node>>) -> Result<Option<String>> {
+        .unwrap()
+        .number()
+        + 1;
+
+    let mut rng = rand::thread_rng();
+
+    let mut id = [0u8; 32];
+    rng.fill(&mut id);
+
+    let id = B256::from_slice(&id);
+
+    node.filters.put(
+        id,
+        (Filter::PendingTransaction, first_block, Instant::now()),
+    );
+
+    Ok(id.to_hex())
+}
+
+fn filter_logs(
+    node: &Node,
+    blocks: impl Iterator<Item = Result<Block>>,
+    address: &Option<OneOrMany<Address>>,
+    topics: &[Option<OneOrMany<B256>>],
+) -> Result<Vec<eth::Log>> {
+    // Get the receipts for each transaction. This is an iterator of (receipt, txn_index, txn_hash, block_number, block_hash).
+    let receipts = blocks
+        .map(|block: Result<_>| {
+            let block = block?;
+            let block_number = block.number();
+            let block_hash = block.hash();
+            let receipts = node.get_transaction_receipts_in_block(block_hash)?;
+
+            Ok(block
+                .transactions
+                .into_iter()
+                .enumerate()
+                .zip(receipts)
+                .map(move |((txn_index, txn_hash), receipt)| {
+                    (receipt, txn_index, txn_hash, block_number, block_hash)
+                }))
+        })
+        .flatten_ok();
+
+    // Get the logs from each receipt and filter them based on the provided parameters. This is an iterator of (log, log_index, txn_index, txn_hash, block_number, block_hash).
+    let logs = receipts
+        .map(|r: Result<_>| {
+            let (receipt, txn_index, txn_hash, block_number, block_hash) = r?;
+            Ok(receipt
+                .logs
+                .into_iter()
+                .filter_map(|l| l.into_evm())
+                .enumerate()
+                .map(move |(i, l)| (l, i, txn_index, txn_hash, block_number, block_hash)))
+        })
+        .flatten_ok()
+        .filter_ok(|(log, _, _, _, _, _)| {
+            address
+                .as_ref()
+                .map_or(true, |address| address.contains(&log.address))
+        })
+        .filter_ok(|(log, _, _, _, _, _)| {
+            topics
+                .iter()
+                .zip(log.topics.iter())
+                .all(|(filter_topic, log_topic)| {
+                    filter_topic
+                        .as_ref()
+                        .map_or(true, |topic| topic.contains(log_topic))
+                    // filter_topic.is_empty() || filter_topic.contains(log_topic)
+                })
+        });
+
+    // Finally convert the iterator to our response format.
+    let logs = logs.map(|l: Result<_>| {
+        let (log, log_index, txn_index, txn_hash, block_number, block_hash) = l?;
+        Ok(eth::Log::new(
+            log,
+            log_index,
+            txn_index,
+            txn_hash,
+            block_number,
+            block_hash,
+        ))
+    });
+
+    logs.collect()
+}
+
+fn get_filter_changes(params: Params, node: &Arc<Mutex<Node>>) -> Result<serde_json::Value> {
```

Comment:
Could consider returning https://docs.rs/alloy/latest/alloy/rpc/types/enum.FilterChanges.html to avoid having to convert to JSON here?